### PR TITLE
Convert path-like strings to `pathlib.Path` objects

### PR DESCRIPTION
**Pull request by happz**
To introduce clear boundary between paths, commands, scripts and the rest of tmt internals.

### Checklist

* [x] implement the feature
* [ ] write documentation
* [ ] extend the test coverage
* [ ] update specification
* [ ] adjust module docs
* [ ] add a usage example
* [ ] modify json schema
* [ ] mention version

### CODE REVIEW

**Comment by psss** [comment]
Let's try to include this one in `1.21` as well.

File: tests/unit/test_utils.py
Comment (psss) [reply tests/unit/test_utils.py:32]: Or shall we use the [tmp_path](https://docs.pytest.org/en/7.1.x/reference/reference.html#tmp-path) instead which returns `pathlib.Path` objects?

**Review by psss — CHANGES_REQUESTED** [review]

File: tests/unit/test_beakerlib.py
Comment (happz) [reply tests/unit/test_beakerlib.py:28]: To make sure we're comparing paths as simplified as possible. Those paths might lead to equal files and directories when followed, but that comparison is harder to make, therefore resorting to comparing them as strings.

```python
>>> Path('/etc/../etc') == Path('/etc')
False
>>> Path('/etc/../etc').resolve() == Path('/etc').resolve()
True
>>> 
```

File: tests/unit/test_utils.py
Comment (happz) [reply tests/unit/test_utils.py:32]: WRT `tmp_path` - that's a possibility, I'd just leave it for another patch, TBH, as I don't know that fixture to just throw it in this patch. I was trying to get rid of path-like strings, to make them easier to follow in the code, and `tmpdir` is not a string, so here it's kind of fine from the perspective of this patch.

Comment (happz) [reply tests/unit/test_utils.py]: `joinpath` => `/` transition happened in 254692fb02bdc62077d8e334624da72ed33e9c6b

File: tests/unit/test_beakerlib.py
Comment (psss) [reply tests/unit/test_beakerlib.py:28]: Hm, ok, I see. I would not expect that the `destination` would contain anything else than an alternative name of the `libs` directory, but we're not doing any checks for that, the [spec](https://tmt.readthedocs.io/en/latest/spec/tests.html#require) also does not limit the value, so your approach seems to be right.

File: tests/unit/test_utils.py
Comment (psss) [reply tests/unit/test_utils.py]: > WRT `tmp_path` - that's a possibility, I'd just leave it for another patch,

Ack, makes sense.

File: tmt/steps/report/html.py
Comment (psss) [inline tmt/steps/report/html.py:77]: Seems a bit complicated. If `pathlib` does not provide an elegant solution for relative paths why not just keep the `os.path.relpath` one-liner as it was?

File: tmt/steps/report/display.py
Comment (psss) [inline tmt/steps/report/display.py:30]:
```suggestion
            # TODO: this should be done already, result.log should use Path instead
```

**Comment by lukaszachy** [comment]
internal-plugins -> should be OK once tmt-redhat MR !68 is squashed and merged
internal-wow  -> MR TBD

File: tmt/steps/discover/__init__.py
Comment (psss) [inline tmt/steps/discover/__init__.py:114]:
```suggestion
            target_dir.mkdir(parents=True, exist_ok=True)
```

We should ensure any intermediate directories are created if needed as well, right?

**Comment by lukaszachy** [comment]
This error message should be reformulated, it cannot be string any more, right? https://github.com/teemtee/tmt/blob/pathlike-strings-to-pathlib-path/tmt/utils.py#L881

File: tmt/steps/discover/shell.py
Comment (psss) [inline tmt/steps/discover/shell.py:344]: Perhaps include a short `TODO` comment?

File: tmt/steps/provision/testcloud.py
Comment (psss) [inline tmt/steps/provision/testcloud.py:693]: Not needed anymore?

File: tmt/steps/report/display.py
Comment (happz) [reply tmt/steps/report/display.py]: Fixed in 3a5a9bacf6ceedddfec98aecab1720f57e28890b

File: tmt/steps/__init__.py
Comment (happz) [reply tmt/steps/__init__.py]: You're right, fixed in 3a5a9bacf6ceedddfec98aecab1720f57e28890b.

File: tmt/steps/discover/shell.py
Comment (happz) [reply tmt/steps/discover/shell.py]: Not needed, actually, I removed those two lines in c91be698.

File: tmt/steps/report/html.py
Comment (happz) [reply tmt/steps/report/html.py]: Trying out the custom `Path`  way in 9275aeabe66a5bcd3e8d57d70bbcb428c9ac2957 - if it works, it seems to me easier to work with.

File: tmt/steps/discover/__init__.py
Comment (happz) [reply tmt/steps/discover/__init__.py]: Correct, fixed.

File: tmt/utils.py
Comment (lukaszachy) [inline tmt/utils.py:919]: This is no longer valid error message

Comment (happz) [reply tmt/utils.py]: Fixed in 1c7cd8dd

Comment (psss) [reply tmt/utils.py]: Thanks.

File: tmt/steps/discover/shell.py
Comment (psss) [reply tmt/steps/discover/shell.py]: Good.

File: tmt/steps/discover/__init__.py
Comment (psss) [reply tmt/steps/discover/__init__.py]: Good.

File: tmt/steps/report/html.py
Comment (psss) [reply tmt/steps/report/html.py]: Yes, this looks much more clear & tidy. Thanks!

File: tmt/convert.py
Comment (thrix) [reply tmt/convert.py:87]: no functional change was obviously not correct, maybe a test should be added to cover the intention?

File: tmt/steps/discover/fmf.py
Comment (psss) [inline tmt/steps/discover/fmf.py:466]: I'm afraid we need to make sure the `path` does not start with a slash (as it is done on line 462), otherwise the last path element overwrites everything.

File: tmt/convert.py
Comment (psss) [reply tmt/convert.py:87]: This part is `nitrate` integration, so test coverage is possible only using `requre`, that is to extend `/tests/integration/the_rest`. Not sure we should block on this.

Comment (happz) [reply tmt/convert.py:87]: That's a mistake, fixed in 040efa1c, hopefully. This `Manual` directory buggers me every time I run `git status`, but this is not the way I plan to get rid of it...

File: tmt/steps/discover/fmf.py
Comment (happz) [reply tmt/steps/discover/fmf.py:466]: Is there a test where I could observe the change? In other words, how did you notice there's an issue?

My current patch would be the following, but I'd like to find out whether it's correct:

```python
prefix_path = Path('/') / 'tests' / (path.relative_to('/') if path.is_absolute() else path)
```

File: tmt/convert.py
Comment (psss) [reply tmt/convert.py:87]: Yeah, now looks good.

File: tmt/steps/discover/fmf.py
Comment (psss) [reply tmt/steps/discover/fmf.py]: Should be fixed and covered by a142cd00.

File: tmt/steps/provision/__init__.py
Comment (psss) [inline tmt/steps/provision/__init__.py:293]: Shouldn't we rather remove the leading slash if present?

Comment (happz) [reply tmt/steps/provision/__init__.py:293]: Hmm, maybe. I don't see the leading slash removed in the original code, `os.path.join` would also fail on this. But I can't dispute your logic.

File: tmt/base.py
Comment (psss) [inline tmt/base.py:368]: The `destination` key should be removed as it's covered below.

File: tmt/steps/provision/__init__.py
Comment (happz) [reply tmt/steps/provision/__init__.py]: Attempting to solve it in d6efd994

File: tmt/base.py
Comment (happz) [reply tmt/base.py]: Right, fixed in 87870ccc

File: tmt/base.py
Comment (psss) [inline tmt/base.py:833]: If the test metadata directory is the same as the `fmf` tree root then we get a `.`, right?

```
>>> pathlib.Path("/a/b/c").relative_to("/a/b/c")
PosixPath('.')
```

So we should not compare with `Path.cwd()`.

File: tmt/base.py
Comment (psss) [inline tmt/base.py:939]:
```suggestion
        # Manual tests should store a path to a document describing the test case steps.
```

File: tmt/steps/provision/__init__.py
Comment (psss) [reply tmt/steps/provision/__init__.py]: Thanks, `Path.unrooted()` will definitely be useful.

File: tmt/base.py
Comment (psss) [reply tmt/base.py]: I guess we need to compare with `Path(".")`, right?

```python
>>> Path('/a/b/c').relative_to(Path('/a/b/c')) == '.'
False

>>> Path('/a/b/c').relative_to(Path('/a/b/c')) == Path('.')
True
```

Comment (happz) [reply tmt/base.py]: I used `if relative_path == '.'` in 9eaf11a, shouldn't it be more or less what used to be in this file before my work?

Comment (psss) [reply tmt/base.py]: Before we were comparing `str` with `str` and now it's `Path` with `str`:

```python
>>> Path('.') == '.'
False
```

Or have I missed anything?

Comment (happz) [reply tmt/base.py]: Ah, now I get it. Fixed in c8c13e04.

Comment (psss) [reply tmt/base.py]: Good, thanks.

File: tmt/base.py
Comment (psss) [inline tmt/base.py:1670]: The `plan_id.path` does not have to start with a slash. Should be fixed by 40675b5c.

File: tmt/steps/discover/fmf.py
Comment (psss) [reply tmt/steps/discover/fmf.py]: Test passes.

File: tmt/utils.py
Comment (happz) [reply tmt/utils.py:919]: Right, `is_dir()` will need to call `stat` system call, which could fail. Makes sense.

Is it supposed to be `in`, wouldn't `.startswith('/')` be better?

File: tmt/utils.py
Comment (happz) [reply tmt/utils.py]: Fixed in e4ba3fdb

Comment (psss) [reply tmt/utils.py]: Good.

File: tmt/utils.py
Comment (happz) [reply tmt/utils.py]: Fixed in dc6bac5a

File: tmt/base.py
Comment (psss) [reply tmt/base.py]: Test passed, looks to be ok.

**Review by psss — APPROVED** [review]
Seems we're done! Ufff! I've just added one last adjustment 36d2ef75 to fix the relative path case for run id. Thanks much for all the adjustments and for the patience!

**Comment by lukaszachy** [comment]
Out-of-the-tree plugins fixed however would be nice to merge only after tests/full is green (running locally).

**Review by lukaszachy — CHANGES_REQUESTED** [review]
tests/full is failing in features/advanced -> /tests/discover/libraries

Simpler steps
```
$ cd tests/discover/libraries/data
tmt run -avv plan -n /plan/strip_git_suffix

total: 3 tests failed
```

**Comment by lukaszachy** [comment]
Thanks, running again in TJ#7541988, TJ#7541986, TJ#7541985, TJ#7541983 (bigger chance  to finish in time when split by plans)

**Comment by lukaszachy** [comment]
Nice, all but /tests/provision/virtual/dependencies (which is VM in VM in VM and failed to boot within 60s) are green.

**Review by lukaszachy — APPROVED** [review]

**Comment by lukaszachy** [comment]
@psss All internal packages are built so IMO we can merge.

**Comment by happz** [comment]
Hmm, there are places that somehow sneaked under the radar, around 20 calls to various `os.path` functions. I'd like to fix them, but given the review & green test, I'd leave that for the follow-up patches...

**Comment by psss** [comment]
> @psss All internal packages are built so IMO we can merge.

Nice, thanks. And thanks also for catching the full test suite issue.

**Comment by psss** [comment]
> Hmm, there are places that somehow sneaked under the radar, around 20 calls to various `os.path` functions. I'd like to fix them, but given the review & green test, I'd leave that for the follow-up patches...

Agreed, let's cover that in a separate pull request.

**Pull request merged**
